### PR TITLE
[FEAT] ICS20-V2 Path Forwarding

### PR DESCRIPTION
**Pull request by chatton**
## Description

Path forwarding feature for ICS20-V2 :rocket:

ref: #6536

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/apps/transfer/keeper/msg_server.go
Comment (chatton) [inline modules/apps/transfer/keeper/msg_server.go:96]: [nit] maybe we should add a comment that the only real reason this could fail, is if exceeding 16 hops by using the unwind field.

File: modules/apps/transfer/keeper/msg_server_test.go
Comment (chatton) [inline modules/apps/transfer/keeper/msg_server_test.go:154]: looks like we mix using `json.Marshal` and `mustMarshalType`

File: CHANGELOG.md
Comment (colin-axner) [reply CHANGELOG.md:76]: Maybe we should have a larger entry to highlight the changes to transfer?

File: modules/apps/callbacks/ibc_middleware_test.go
Comment (colin-axner) [reply modules/apps/callbacks/ibc_middleware_test.go:27]: I like that idea

File: modules/apps/transfer/ibc_module.go
Comment (colin-axner) [reply modules/apps/transfer/ibc_module.go:216]: I like it on `msgTransfer`, but maybe should be different on the packet data. `data.IsForwardingPacket()`? Something like that?

File: modules/apps/transfer/internal/events/events.go
Comment (colin-axner) [reply modules/apps/transfer/internal/events/events.go:15]: question for @DimitrisJim

File: modules/apps/transfer/keeper/keeper.go
Comment (colin-axner) [inline modules/apps/transfer/keeper/keeper.go:319]: This should be private and a grpc should be added instead?

File: modules/apps/transfer/types/keys.go
Comment (colin-axner) [inline modules/apps/transfer/types/keys.go:33]: unused, delete

File: modules/apps/transfer/types/keys.go
Comment (colin-axner) [inline modules/apps/transfer/types/keys.go:37]: a bit of inconsistency we have here. Some keys are defined below as consts with a byte prefix, while other keys (total escrow and this packet forward prefix) use string prefixes. The byte prefixes use a prefix store in the keeper setting code while the string prefixes use helper functions. In general would prefer store prefixes to be grouped together in one section so it is easy to understand what keys exist for state storage

File: modules/apps/transfer/client/cli/tx.go
Comment (crodriguezvega) [reply modules/apps/transfer/client/cli/tx.go:159]: I guess if the port/channel pairs are messed up, then it would fail in the message validation. But maybe it's worth also running here the validation for port ID and channel ID.

File: modules/apps/transfer/ibc_module.go
Comment (damiannolan) [reply modules/apps/transfer/ibc_module.go:216]: Think I'd vote for `HasFowarding` :D

File: modules/apps/transfer/keeper/relay.go
Comment (crodriguezvega) [inline modules/apps/transfer/keeper/relay.go:493]: Looks like this function can be removed and we can call `bankKeeper.BurnCoins` directly, right?

Comment (crodriguezvega) [reply modules/apps/transfer/keeper/relay.go]: https://github.com/cosmos/ibc-go/pull/6693/commits/cc46a63ea6e0644082b25761efbef0abfd759d91

File: modules/apps/transfer/keeper/keeper.go
Comment (crodriguezvega) [reply modules/apps/transfer/keeper/keeper.go]: I made it private, but didn't add a grpc. Maybe we can add it later if the usage for it arises?

File: modules/apps/transfer/ibc_module.go
Comment (damiannolan) [reply modules/apps/transfer/ibc_module.go]: Whoops! I pushed a commit changing both to `HasFowarding` for consistency. If you guys feel strongly we can change, either `Is` or `Has` is fine by me for dump bool return funcs, just prefer to follow that convention

File: modules/apps/callbacks/ibc_middleware_test.go
Comment (damiannolan) [reply modules/apps/callbacks/ibc_middleware_test.go:27]: Same for `Forwarding` going into the `MsgTransfer`, looks like a lot of `transfertypes.Forwarding{}` diffs here! There's a good bit of that littered around now.

I would also suggest that I'd be fine with having a `WithForwarding(forwarding Forwarding)` method on the msg or packet data for assigning the forwarding, since it is an optional feature. Would be curious to hear what others think. This is obviously something that can be dealt with post-merge of this PR too. Would be happy stab the refactor myself 🗡️

File: modules/apps/transfer/client/cli/tx.go
Comment (damiannolan) [reply modules/apps/transfer/client/cli/tx.go:47]: Another one to add to your suggestion @DimitrisJim 

Please note that if the {unwind} flag _is_ used

File: modules/apps/callbacks/ibc_middleware_test.go
Comment (chatton) [reply modules/apps/callbacks/ibc_middleware_test.go:27]: think I would prefer to keep the message signature the same and have it as a required field during construction, but have an obvious "I don't care about this" value.

File: modules/apps/transfer/client/cli/tx.go
Comment (damiannolan) [reply modules/apps/transfer/client/cli/tx.go:155]: I think you'd need to use a different default instead of `""` then, right? in order to be able to return an error on empty string if the flag is there. Unless there's a cuter way to do it

File: modules/apps/transfer/ibc_module.go
Comment (damiannolan) [inline modules/apps/transfer/ibc_module.go:181]:
```suggestion
// A nil acknowledgement may be returned when using the packet forwarding feature. This signals to core IBC that the acknowledgement will be written asynchronously.
```

File: modules/apps/transfer/client/cli/tx.go
Comment (damiannolan) [reply modules/apps/transfer/client/cli/tx.go:159]: channel IDs shouldn't have slashes in them afaik

`var IsChannelIDFormat = regexp.MustCompile(`^channel-[0-9]{1,20}$`).MatchString`

File: modules/apps/transfer/keeper/keeper.go
Comment (colin-axner) [reply modules/apps/transfer/keeper/keeper.go]: yea sounds good!

File: modules/apps/transfer/keeper/forwarding.go
Comment (colin-axner) [reply modules/apps/transfer/keeper/forwarding.go:1]: you are the greatest for combing through code cov!! :heart: :1st_place_medal:

File: modules/apps/transfer/keeper/relay.go
Comment (AdityaSripal) [inline modules/apps/transfer/keeper/relay.go:303]: this doesn't need to be a helper fn

File: modules/apps/transfer/keeper/relay.go
Comment (AdityaSripal) [inline modules/apps/transfer/keeper/relay.go:315]: this doesn't need to be helper fn

File: modules/apps/transfer/keeper/relay.go
Comment (AdityaSripal) [inline modules/apps/transfer/keeper/relay.go:334]: ditto

File: modules/apps/transfer/keeper/relay.go
Comment (AdityaSripal) [inline modules/apps/transfer/keeper/relay.go]: does this helper function need to exist?

File: modules/apps/transfer/types/forwarding.go
Comment (AdityaSripal) [inline modules/apps/transfer/types/forwarding.go:9]: I think we should make this 8 personally

**Review by AdityaSripal — APPROVED** [review]
Reviewed the non-test files and acking for merge.

I do have some nits on structuring. In general, I found the use of helper functions too extensive, to the point where it hurt understandability more than it helped. Left comments where relevant. This can be discussed and fixed in later PRs

Acking to unblock since i mainly wanted to review the overall logic in key places. I can review test files as well another day if folks want me to

File: modules/apps/transfer/client/cli/tx.go
Comment (chatton) [reply modules/apps/transfer/client/cli/tx.go:159]: ah sorry I was mixing it up with denoms which can have slashes, we should be fine with this

File: modules/apps/transfer/keeper/msg_server.go
Comment (chatton) [reply modules/apps/transfer/keeper/msg_server.go:43]: guess we could do the unwind + lots of hops to hit an error in validate basic! Nice catch

File: modules/apps/transfer/types/transfer_authorization.go
Comment (crodriguezvega) [reply modules/apps/transfer/types/transfer_authorization.go:1]: Good catch. We should add it to `ValidateBasic()`. I can do it if you want.

File: modules/apps/transfer/internal/convert/convert.go
Comment (crodriguezvega) [reply modules/apps/transfer/internal/convert/convert.go:11]: This function is actually not used and we have the same one now in `packet.go`, so I will remove it.

File: e2e/testsuite/testsuite.go
Comment (chatton) [reply e2e/testsuite/testsuite.go:254]: there's actually a small restructure that needs to happen here #6721 I will address this in that PR once it gets synced with main after this is merged

File: modules/apps/transfer/keeper/relay.go
Comment (colin-axner) [reply modules/apps/transfer/keeper/relay.go]: I think the todo can be addressed now as a side note

**Review by colin-axner — APPROVED** [review]
Woo!! Reviewed the core logic and left any relevant notes. I skimmed over the tests, but largely relying on the fact that we are checking code cov and have already reviewed all the files changed. Happy to see this merged and followup issues addressed after the fact

File: modules/apps/transfer/internal/convert/convert.go
Comment (crodriguezvega) [reply modules/apps/transfer/internal/convert/convert.go]: https://github.com/cosmos/ibc-go/pull/6693/commits/7d2a881f0a92b194220ccbce0cbf9bf45fd82b84

File: modules/apps/transfer/keeper/relay.go
Comment (crodriguezvega) [reply modules/apps/transfer/keeper/relay.go]: https://github.com/cosmos/ibc-go/pull/6693/commits/4ea36c22a574a29743d4084e0927973edea30009

File: modules/apps/transfer/keeper/relay.go
Comment (damiannolan) [reply modules/apps/transfer/keeper/relay.go:303]: I agree with the general sentiment. I also find there's an excessive use of smaller funcs too

Comment (crodriguezvega) [reply modules/apps/transfer/keeper/relay.go:303]: For context: the pros and cons of having these helper functions were discussed [here](https://github.com/cosmos/ibc-go/pull/6574#discussion_r1636013788).

File: modules/apps/transfer/keeper/msg_server.go
Comment (damiannolan) [inline modules/apps/transfer/keeper/msg_server.go:73]: nit: there's no need to return the msg when we're mutating a pointer, right?

func sig could just return error

File: modules/apps/transfer/keeper/relay.go
Comment (damiannolan) [inline modules/apps/transfer/keeper/relay.go:324]: nit: I'd vote for godocs to specify ordinary transfer logic first, and then state what happens if its a multihop transfer

can clean these up post merge

File: modules/apps/transfer/types/keys.go
Comment (damiannolan) [reply modules/apps/transfer/types/keys.go:36]: Let's group them together in a follow up PR

File: modules/apps/transfer/types/transfer_authorization.go
Comment (damiannolan) [inline modules/apps/transfer/types/transfer_authorization.go:179]: should we `Wrapf` the unauthorized hops?

**Review by damiannolan — APPROVED** [review]
LGTM! Thanks everyone for their work on this feature! ❤️

File: modules/apps/transfer/keeper/relay.go
Comment (damiannolan) [reply modules/apps/transfer/keeper/relay.go:324]: created https://github.com/cosmos/ibc-go/issues/6724

File: modules/apps/transfer/types/transfer_authorization.go
Comment (damiannolan) [reply modules/apps/transfer/types/transfer_authorization.go]: https://github.com/cosmos/ibc-go/pull/6693/commits/7411f2d16e9cce64c4ecf09c201e4ed5b8aaab5d

File: modules/apps/callbacks/ibc_middleware_test.go
Comment (crodriguezvega) [reply modules/apps/callbacks/ibc_middleware_test.go:1]: I will add this to the e2e test scenarios in https://github.com/cosmos/ibc-go/issues/6578.

File: modules/apps/transfer/keeper/msg_server.go
Comment (chatton) [reply modules/apps/transfer/keeper/msg_server.go:73]: I think I'd prefer to not modify `msg` and return a new one entirely. Not a huge deal though

**Review by chatton — COMMENTED** [review]
Can't leave an approval since I created the PR but I can leave a :green_circle: .

File: modules/apps/transfer/types/transfer_authorization.go
Comment (crodriguezvega) [reply modules/apps/transfer/types/transfer_authorization.go:1]: https://github.com/cosmos/ibc-go/pull/6693/commits/e310e27419dac240a184f4d26c6ec52d701f55d5

File: modules/apps/transfer/keeper/msg_server.go
Comment (crodriguezvega) [reply modules/apps/transfer/keeper/msg_server.go:43]: https://github.com/cosmos/ibc-go/pull/6693/commits/f28e9374b029b6678b354f9fef21d0b43d39134c

**Comment by crodriguezvega** [comment]
I am looking at the upgrade tests that fail to see if I can do a quick fix before merging.

**Pull request merged**
